src/systems/filecoin_blockchain/storage_power_consensus/storage_power_actor_code.go

```diff
@@ -91,13 +93,14 @@ func (a *StoragePowerActorCode_I) WithdrawBalance(rt Runtime, amount actor.Token
 
 	h, st := a.State(rt)
 
-	currEntry, found := st.PowerTable()[minerID]
-	if !found {
-		rt.Abort("minerID not found.")
+	ret := st._ensurePledgeCollateralSatisfied(rt)
+	if !ret {
+		rt.Abort("insufficient pledge collateral.")
 	}
 
+	currEntry := st._safeGetPowerEntry(rt, minerID)
 	if currEntry.AvailableBalance() < amount {
```

Comment:
I'm a bit concerned about manual balance accounting -- I think, as much as possible, we should statically enforce that currency is not created or destroyed in any of these operations, but rather is only moved via Send calls (or `_transferFunds` calls, if needed outside the VM).

In this particular case, instead of `AvailableBalance`, I would recommend that the singleton actors each keep a simple running total, per sender-actor, of how much currency they are currently holding on that sender's behalf. Then it's easy to abstract this out, so that we have a single method that ensures this is accurate, rather than custom logic to do the accounting separately for each singleton actor. Separately, the particular actor in question (here StoragePowerActor) can then do collateral arithmetic on top of that running total, to determine whether pledge collateral requirements are satisfied.

Reply:
Just to be on the same page, the sum of AvailableBalance and LockedBalance is the running total. You are saying we should just check if pledge collateral is satisfied as a function on the total without keeping tracking of Avail vs Locked and we will do this collateral check before withdraw balance as well.

Reply:
Yes -- essentially, using a single tally in StoragePowerActor (or more precisely, a map from actor to tally), rather than using two separate tallies for Available and Locked. I think it's a relatively small change, but conceptually much less error-prone, because if singleton actors in general have a single mechanism for maintaining escrow funds on behalf of individual actors, then we can reuse the same machinery everywhere.

Reply:
What was the initial purpose of the distinction @zixuanzh ?

Reply:
@sternhenri this is to accommodate pledge collateral changing dynamically and miners may switch between over and under collateralization quite frequently. Available balance is acting as a buffer or otherwise SPA needs to send and receive collateral very frequently.